### PR TITLE
[FLINK-15854][hive] Use the new type inference for Hive UDTF

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
Use new type inference for Hive UDTF

## Brief change log
  -  Stop creating `TableFunctionDefinition` in `HiveFunctionDefinitionFactory`
  -  Remove `HiveTableSqlFunction`
  -  Add `HiveFunctionArguments` to wrap all argument information needed by a hive function
  -  Support NULL literal arguments
  -  Add new `HiveFunction` interface and move the type inference logic to this class
  -  Convert string literal to string type 

## Verifying this change
Existing tests and added test for udtf `json_tuple` in  `TableEnvHiveConnectorITCase#testUDTF `.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by luoyuxia** [comment]
It's a rework on #13341

**Comment by luoyuxia** [comment]
@JingsongLi @twalthr Could you please help review?

**Comment by luoyuxia** [comment]
@flinkbot run azure

**Comment by twalthr** [comment]
I also worked on the function stack and found some shortcomings in how we translate to Calcite. I would like to merge FLINK-26518 before this one. But I haven't forgotten about this one.

**Comment by luoyuxia** [comment]
@twalthr Thanks for your attention.

**Comment by twalthr** [comment]
The other PR is merged. Could you rebase to check whether everything still works? Will do a full review shortly.

**Comment by luoyuxia** [comment]
@twalthr Thanks for your reminder. I rebased now.

**Comment by luoyuxia** [comment]
@twalthr Sorry to bother you, but do you have time to review this pr? Then I can continue the new type inference for Hive UDAF.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveFunction.java [hunk 1]
```diff
@@ -19,38 +19,104 @@
...
-     * @param argTypes types of arguments
-     * @return result type.
+     * @return The inferred return type.
+     * @throws UDFArgumentException can be thrown if the input arguments are invalid.
      */
-    DataType getHiveResultType(Object[] constantArguments, DataType[] argTypes);
+    DataType inferReturnType() throws UDFArgumentException;
+
+    /** Gets the wrapper for the Hive function. */
+    HiveFunctionWrapper getFunctionWrapper();
```
Comment (wuchong) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveFunction.java:55]: Declare the generic type `<UDFType>` in interface?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserCalcitePlanner.java [hunk 2]
```diff
@@ -2539,9 +2539,9 @@ private RelNode genUDTFPlan(
 
         SqlOperator convertedOperator = convertedCall.getOperator();
         Preconditions.checkState(
-                convertedOperator instanceof SqlUserDefinedTableFunction,
+                convertedOperator instanceof BridgingSqlFunction,
                 "Expect operator to be "
-                        + SqlUserDefinedTableFunction.class.getSimpleName()
+                        + BridgingSqlFunction.class.getSimpleName()
```
Comment (wuchong) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserCalcitePlanner.java:2544]: What's the reason for changing this?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/TableEnvHiveConnectorITCase.java [hunk 3]
```diff
@@ -288,6 +290,15 @@ public void testUDTF() throws Exception {
                                     .collect());
             assertThat(results.toString()).isEqualTo("[+I[{1=a, 2=b}], +I[{3=c}]]");
 
+            assertThat(results.toString()).isEqualTo("[+I[{1=a, 2=b}], +I[{3=c}]]");
```
Comment (wuchong) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/TableEnvHiveConnectorITCase.java:293]: This assertion is redundant?

**Review by wuchong — COMMENTED** [review]
Thanks for the contribution @luoyuxia , it's great to see we are getting rid of the old stack. 

I have left some small comments.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserCalcitePlanner.java (hunk 2 above)
Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserCalcitePlanner.java:2544]: Before this pr, the Hive's udtf will be converted to `HiveTableSqlFunction`, which is instance of SqlUserDefinedTableFunction. The old code logic is in  [here](https://github.com/apache/flink/pull/18958/files#diff-ca5db8275fd03a414acc2af23a9114f47791506fb0542986f55f31865f1dbb87L130).
After this pr, it'll then convert to `BridgingSqlFunction`.
So, we need to.change it.

**Comment by luoyuxia** [comment]
@wuchong We haven't enabled the CI test for Hive 3.x. Usually I wiil run the test in my local  manually for Hive 3.x with command `mvn  test -Phive-3.1.1`

**Comment by luoyuxia** [comment]
@flinkbot run azure

**Comment by wuchong** [comment]
> @wuchong We haven't enabled the CI test for Hive 3.x. Usually I wiil run the test in my local manually for Hive 3.x with command `mvn test -Phive-3.1.1`

Could you open an issue to enable the hive-3.1.1 test in the pre-commit? I think it's hard to guarantee the compatibility for Hive 3.x by manually testing.

**Review by wuchong — APPROVED** [review]
LGTM.

**Comment by luoyuxia** [comment]
@wuchong I have opened the issue [FLINK-27895](https://issues.apache.org/jira/browse/FLINK-27895).

**Pull request merged**